File: sdks/typescript/octollm-sdk/examples/basicUsage.ts
/**
 * Basic usage example for OctoLLM SDK
 *
 * Demonstrates submitting a task and polling for completion.
 */

import { OrchestratorClient, TaskRequest } from '../src';

async function main() {
  // Initialize client with API key
  const client = new OrchestratorClient({
    baseUrl: 'http://localhost:8000',
    apiKey: process.env.OCTOLLM_API_KEY || 'your-api-key-here',
    timeout: 30000
  });

  // Create a task request
  const taskRequest: TaskRequest = {
    goal: 'Analyze the security vulnerabilities in the provided Python web application',
    constraints: [
      'Focus on OWASP Top 10 vulnerabilities',
      'Include SQL injection and XSS analysis'
    ],
    acceptance_criteria: [
      'All high-severity vulnerabilities identified',
      'Remediation recommendations provided',
      'Code examples for fixes included'
    ],
    context: {
      framework: 'Flask',
      python_version: '3.11',
      authentication: 'JWT'
    },
    budget: {
      max_tokens: 10000,
      max_time_seconds: 120,
      max_cost_dollars: 1.0
    }
  };

  try {
    // Submit task
    console.log('Submitting task...');
    const response = await client.submitTask(taskRequest);
    console.log(`Task submitted: ${response.task_id}`);
    console.log(`Status: ${response.status}`);
    console.log(`Created at: ${response.created_at}`);

    // Poll for completion
    console.log('\nPolling for completion...');
    let attempts = 0;
    const maxAttempts = 60; // 2 minutes with 2-second intervals

    while (attempts < maxAttempts) {
      await new Promise(resolve => setTimeout(resolve, 2000)); // Wait 2 seconds

      const status = await client.getTask(response.task_id);
      console.log(`Status: ${status.status}`);

      if (status.progress) {
        console.log(
          `Progress: ${status.progress.percentage}% ` +
          `(${status.progress.current_step})`
        );
      }

      // Check if task is complete
      if (['completed', 'failed', 'cancelled'].includes(status.status)) {
        console.log('\nTask finished!');

        if (status.status === 'completed' && status.result) {
          console.log('\n=== RESULT ===');
          console.log(`Output: ${status.result.output}`);
          console.log(`Confidence: ${status.result.confidence}`);
          console.log(`Validation Passed: ${status.result.validation_passed}`);
        } else if (status.status === 'failed' && status.error) {
          console.log('\n=== ERROR ===');
          console.log(`Type: ${status.error.type}`);
          console.log(`Message: ${status.error.message}`);
          if (status.error.details) {
            console.log(`Details: ${status.error.details}`);
          }
        }

        break;
      }

      attempts++;
    }

    if (attempts >= maxAttempts) {
      console.log('\nTimeout: Task did not complete in time');
    }

  } catch (error: any) {
    console.error('Error:', error.message);
    if (error.statusCode) {
      console.error(`Status Code: ${error.statusCode}`);
    }
    if (error.details) {
      console.error('Details:', error.details);
    }
  }
}

// Run the example
main().catch(console.error);

File: sdks/python/octollm-sdk/examples/basic_usage.py
"""
Basic usage example for OctoLLM Python SDK.

This example demonstrates:
- Initializing the orchestrator client
- Submitting a task
- Polling for task completion
- Handling results
"""

import asyncio
from octollm_sdk import (
    OrchestratorClient,
    TaskRequest,
    ResourceBudget,
    OctoLLMError,
)


async def main():
    # Initialize client with API key
    # You can also set OCTOLLM_API_KEY environment variable
    client = OrchestratorClient(
        base_url="http://localhost:8000",
        api_key="sk-your-api-key-here",  # Replace with your actual API key
        timeout=60.0,  # 60 second timeout
    )

    # Check service health
    print("Checking service health...")
    health = await client.health()
    print(f"Service status: {health.status}")
    print(f"Version: {health.version}")
    print()

    # Create a task request
    task = TaskRequest(
        goal="Create a Python function to validate email addresses using regex",
        constraints=[
            "Include type hints",
            "Add comprehensive docstring",
            "Handle edge cases (empty string, multiple @, etc.)",
        ],
        acceptance_criteria=[
            "Function validates RFC 5322 email format",
            "Function includes at least 3 test cases",
            "Code follows PEP 8 style guide",
        ],
        budget=ResourceBudget(
            max_tokens=5000,
            max_time_seconds=30,
            max_cost_dollars=0.10,
        ),
    )

    # Submit task
    print("Submitting task...")
    print(f"Goal: {task.goal}")
    try:
        response = await client.submit_task(task)
        print(f"Task submitted successfully!")
        print(f"Task ID: {response.task_id}")
        print(f"Status: {response.status}")
        print()
    except OctoLLMError as e:
        print(f"Error submitting task: {e}")
        return

    # Poll for completion
    print("Waiting for task completion...")
    max_attempts = 30
    attempt = 0

    while attempt < max_attempts:
        try:
            status = await client.get_task(response.task_id)

            if status.status == "completed":
                print("\nTask completed successfully!")
                if status.result:
                    print(f"Confidence: {status.result.confidence:.2%}")
                    print(f"Validation passed: {status.result.validation_passed}")
                    print("\nGenerated code:")
                    print("=" * 80)
                    print(status.result.output)
                    print("=" * 80)

                if status.metadata:
                    print(f"\nMetadata:")
                    print(f"  Arms used: {', '.join(status.metadata.arms_used)}")
                    print(f"  Tokens used: {status.metadata.tokens_used}")
                    print(f"  Cost: ${status.metadata.cost_dollars:.4f}")
                    print(f"  Duration: {status.metadata.duration_seconds:.2f}s")
                break

            elif status.status == "failed":
                print("\nTask failed!")
                if status.error:
                    print(f"Error type: {status.error.type}")
                    print(f"Error message: {status.error.message}")
                    if status.error.details:
                        print(f"Details: {status.error.details}")
                break

            elif status.status == "processing":
                if status.progress:
                    print(
                        f"Progress: {status.progress.percentage}% "
                        f"(step {status.progress.completed_steps}/{status.progress.total_steps}: "
                        f"{status.progress.current_step})"
                    )

            # Wait before next poll
            await asyncio.sleep(2)
            attempt += 1

        except OctoLLMError as e:
            print(f"Error checking task status: {e}")
            break

    if attempt >= max_attempts:
        print(f"\nTimeout: Task did not complete within {max_attempts * 2} seconds")


if __name__ == "__main__":
    asyncio.run(main())
